Add context-based station economic industries

sturnclaw opened the pull request:
In short, this PR finishes the placeholder economic simulation that was introduced in #5474. It replaces the three-factor random roll which determined whether a commodity was imported or exported at a station with a set of context-dependent starport industries which consume and produce goods, adds a local "trade neighborhood" modelling short-hop NPC trading between stations around the same primary body, and models long-term oversaturation of commodity trades at markets.

**Please consider this PR an open call for content!**

I've provided a basic set of starter industries, but this is by no means a complete set and I would welcome contributions! New industries do not have to have a wholly unique set of commodities; additional variants of existing industries tailored for unique situations (e.g. high-temperature volcanic worlds might have a special metal smelting industry?) are strongly encouraged.

At current, the PR is "feature complete" however there are some aspects I'd like to explore adding before final merge:
- [ ] ~~Add a neighborhood context for condition tags (e.g. propagate the gas giant tag to starports on its moons).~~ Not used at this time. No current use, introduces hierarchical constraints.
- [ ] ~~Add govtype condition tags to system-wide context. (This would allow for government or economy-specific industries, e.g. Collective Industry for communist govtypes instead of Civil Industry.)~~ Not used at this time. Trivial to add.
- [x] Add a more intelligent system for selecting starport industries. ~~Current thought is a set of "upgrade trees" which are predicated on condition tags (e.g. build ore mine, then build refinery, then build heavy industry). Starport industries would be selected from the valid set of upgrade trees.~~ Implemented as per-industry build-next list, with condition tag and random-chance support.
- [x] Balancing existing commodity prices and quantities.
- [x] Adding additional commodities to round out production chains and create more rational supply/demand networks. (e.g. Rare Ores, Luxury Goods, *medical precursors that aren't food*, etc...)
- [ ] ~~Displaying industries present at a station in the map screens and on the station's comms page.~~ Not relevant at this time.
- [x] Write a full system description for developers in codedoc.
- [ ] Write a full system description for modders / content authors in devdocs.

The condition tags system should be somewhat self-documenting; it works by comparing named fields of the `SystemBody` object to a literal number or string. A few additional condition tags are directly injected into the context evaluated by industries so as to reduce the number of custom-authored condition definitions. An additional documentation pass is probably required on the conditions system, as Conditions.lua is somewhat lacking in documentation.

sturnclaw committed:
Document SystemBody TODO

sturnclaw committed:
Expose StarSystem.seed to Lua

sturnclaw committed:
economy: tweak commodity prices

- Provide more consistent profit margins and station stocking of low-value commodities.

sturnclaw committed:
economy: add Conditions, data-driven named boolean tags

- Conditions allow expressing body attributes as named boolean values for efficient selection/restriction of industries.
- Conditions represent a data-driven way to associate semantic names with specific value thresholds for bodies and systems.
- Conditions represent "local conditions" at a specific starport, and are chained together, combining system -> parent planet -> starport conditions.

sturnclaw committed:
economy: add Industries, starport economic producers

- An Industry is a representative of one or more economic producers which present at a starport, and are the primary mechanism by which starports are differentiated from each other.
- Each Industry has a set of commodity input values which it consumes in the course of its work, and a set of outputs which it likewise produces.
- Input/output values are specified in a non-physical "flow" value representing orders of magnitude over an unspecified time unit.
- Starport economies are based on the aggregate of all Industries present at their location, plus the effects of their Population, a special Industry.

sturnclaw committed:
Lua: add path value to FS.ReadDirectory

- Returns the path relative to the file source (without the protocol part) for use with existing file APIs which take a legacy path.

sturnclaw committed:
Lua: properly document Rand intervals

sturnclaw committed:
economy: add condition and industry definitions

-  l10n_key values for industries are not currently used and do not have a language resource.
- workforce values for industries are not currently used.
- population.json is a special definition which defines the commodity inputs/outputs for a size 1 starport.
- Basic commodity coverage has been provided, but the list of industries is by no means exhaustive.

sturnclaw committed:
economy: add industry-based starport economies

- Industrial supply and demand is used to inform commodity stock availability and price.
- Local supply networks model realistic saturation of local trade, yielding economic opportunitities for the player to import goods from other networks.
- Flow units are used to express commodity trade over an indeterminate standard time period, and are related to instantaneous stock quantities by powers of e.
- Add a debug interface for economic conditions showing domestic supply/demand numbers, local supply network effects, and stock equilibrium and price values.

sturnclaw committed:
ui: fix ui.Number not displaying sign on thousands-place numbers

sturnclaw committed:
economy: add economic simulation using random walks

- Each station tracks supply, demand, and current stock levels for each commodity.
- These levels are constantly changing at a configurable rate, yielding a market that feels "alive" yet remains close to equilibrium over a sufficiently long time period
- The current supply and demand values influence commodity pricing, perturbing the profitability of trades slightly over time.
- The system is intended to support injecting external events which affect supply/demand/stock levels.
- The first such external event is the player's "history register" which records buy/sell amounts per-commodity and perturbs the price of goods over a long timescale.
  This behavior is intended to encourage the player to run multiple different trade routes, and prevents trade gameplay from devolving to a singular maximally-profitable route.

sturnclaw committed:
economy: replace old economy implementation

- Differentiate between persistent and transient commodity markets; persistent markets receive updates and are serialized in the savefile.
- Transient markets are intended for generating data for distant systems, for use by e.g. the trade computer.
- Implement market serialization, including an optimization that saves ~50kb in a Sol start.
- All markets in the current system are stored as persistent markets and receive updates at runtime.
- The commodity market widget and commodity price debug screen have not yet received updates for the new economy semantics.

sturnclaw committed:
Update commodity price debug tab

sturnclaw committed:
Update NewsEventCommodity for new Economy code

- Modify supply/demand values rather than commodity stocks directly.

sturnclaw committed:
economy: remove remaining old-econ code

sturnclaw committed:
CommodityMarket: remove demand, get pricemod from Economy

- The economic simulation tracks demand as an instantaneous rate, rather than a discrete quantity.
  Consequently, there is no longer a demand limit on how many goods the player can sell to the market; instead, the player's actions are recorded to persistent history and influence the commodity price.

sturnclaw committed:
economy: add industries to produce medicines/precious_metals

- Also fix mis-spellings of the medicines commodity ID.

sturnclaw committed:
Remove SpaceStation:GetCommodityDemand

- The prior concept of demand as an instantaneous quantity has been removed and replaced with a rate value; this means demand is no longer needed as a SpaceStation API.

mwerle reviewed (commented):
Initial quick look, no playtesting yet.

Some questions, a few nitpicks. Overall LGTM (as far as I understand what's going on, which isn't very). Would appreciate some more detailed docs of how this is all supposed to work so reviews can compare against the design.

Bodasey commented:
Hard to contribute something without visualisation...but let's try some thoughts, hoping they are what you are looking for.

> * Adding additional commodities to round out production chains

We can make a commodity of every chemical element and base chemical compound (CH4, NH3, H2O, CO2, SiO2, FeO2, ...) depending on how detailed we want to model space mining and trade. E.g. "fertilizer" yet is a tailor made mixture of mainly Nitrogen, Phosphor, Kalium and other elements for different plants and environments.


Condition tags? Ok, let's list some natural resources on the bodies for the surface, crust, hydrosphere and atmosphere. 

> * Add a neighborhood context for condition tags (e.g. propagate the gas giant tag to starports on its moons).

Can be made, but even Hydrogen needs to be scooped and transported, and what about Venus and her CO2? => extended scoop equipment and even scoop missions?

Condition tags have to include type and intensity of energy sources (electric current from magnetic fields, wind, geothermal energy, solar radiation).

> * Add a more intelligent system for selecting starport industries.

Ok, I appreciate your model as intelligent enough if it generates a starport in the orbit around Venus who processes CO2 from her atmosphere with H2 from the outer solar system to Carbon ore and plastics with O2 as byproduct, or even better imports H2O from ice moons to get the H2 from fuel cells because you can directly use electric current from more than 2000W/m2 solar radiation which is far more efficient than the 1340W/m2 we get in Earth's orbit. O2 may then be shipped back to the gas giants to be used in fuel cells, which could be another additional commodity. And oh yes, fuel cells are a source of energy to be tagged if no other source is available, meaning a demand for H2 and O2.

(Airships would be a better solution to scoop gas for permanent industrial supply, shall we model them as orbiting objects?)

sturnclaw commented on data/economy/conditions/basic.json:
```diff
@@ -0,0 +1,78 @@
+		"context": "planet",
+		"conditions": [ "gravity <= 3.3" ]
+	},
+	"low_gravity": {
+		"context": "planet",
+		"conditions": [ "gravity > 3.3" ]
+	},
+	"norm_gravity": {
+		"context": "planet",
+		"conditions": [ "gravity > 6.0" ]
+	},
```
1. Some condition tags are intended as a "at least X" semantic. In this case, the `low_gravity` condition could/should be expressed as `min_low_gravity`, as the condition is intended to be used to select only bodies with at minimum 3.3m/s of surface gravity.
2. The unit is / should be specified in `SystemBody.h` and `data/meta/SystemBody.lua`. The explicit context that a condition definition is evaluated against is a SystemBody object, and the names here are the named fields of a SystemBody.
3. Condition tags were added during development to support industries added during development. The system is designed to be extremely extensible, so no attempt was made at defining a completely rigorous set of conditions "ahead of time".
4. There is no way to OR condition *requirements* together; condition tags represent "named facts" about a body, so all parts of a "fact" must be present for that "fact" to be true. The OR operator is instead expressed via the `conditions_any` field of an industry.

I'm not entirely opposed to adding an OR operator in there somewhere, but I'd need to see a convincing case that isn't expressible via the existing system to justify the added complexity of condition definitions, parsing, and evaluation.

sturnclaw replied:
1. Yes, `(min_)low_gravity` (intentionally) includes all bodies with at least 3.3m/s of surface gravity. If an industry should be limited to *only* 3.3-6.0m/s of surface gravity, the industry's `conditions` field could be something like `[ "low_gravity", "!norm_gravity" ]`. Alternatively, if the intent is to express a constraint that an industry can only exist in a low gravitational field or no gravity at all, the `conditions` field of an industry definition could instead require `[ "!norm_gravity" ]`.

    The philosophy here is that **conditions exist to serve industries**. If a constraint is not correctly expressing the semantic requirements of the industry definitions we have, it should be rewritten to either have different numeric values, a different name, or an entirely different condition added.

    As to developers understanding conditions - the list of conditions is two screens long, written in the most self-documenting way possible. Between *reading the list of conditions* and the documentation page I intend to author explaining the entire system for the use of modders/developers, I do not think there will be any confusion at all. Comments can certainly be added (just add a `"//": "..."` field in the JSON) but are intended to be unneeded as much as possible.

2. A comment where? I'm planning on a full documentation pass *most likely* in the [codedoc](https://codedoc.pioneerspacesim.net) but possibly also in the dev docs. Said documentation would fully explain the system, with links to the relevant documentation pages for allowed fields, etc.

mwerle replied:
1. Ok; I'm really sorry, but I find this set of conditions for gravity really confusing, especially when the conditions for atmospheres just below have specific bands. And needing to combine gravity conditions like that in an industry seems to me to also be unintuitive when the same isn't needed for atmospheres. I most certainly did not find it intuitive or self-documenting. But you can put it down to me being a numpty.
2. I was meaning a comment in the source. That's what people mostly read. But I'm sorry, I forgot this is `JSON` and doesn't natively support comments. One reason I somewhat dislike `JSON`. Having good docs (and reminding people to read them) is the next best thing. Closing.

sturnclaw commented on data/economy/conditions/basic.json:
```diff
@@ -0,0 +1,78 @@
+		"context": "planet",
+		"conditions": [ "atmosDensity >= 1.0", "atmosOxidizing <= 0.7" ]
+	},
+	"atmos_breathable": {
+		"context": "planet",
+		"conditions": [ "atmosDensity >= 1.0", "atmosOxidizing > 0.7" ]
+	},
```
> Is the "atmosDensity" in units of bar?

No, Bar/Atm is a unit of pressure. Atmospheric density is measured in kg/m3, with earth having a 1.225kg/m3 density.

...yes, this is probably the wrong field to use to determine a breathable atmosphere. Suggestions for a minimum and maximum `surfacePressure` value for breathable atmosphere thresholds would be appreciated!

mwerle replied:
As I said, for humans, "breathable" is a combination of surface pressure _and_ oxygen content (and absence of other toxic gases; inert gases don't matter). So might be a bit more complex..

Can the condition system generate derived units / have formulae as part of it? And/or call functions do so for it?

sturnclaw replied:
> Can the condition system generate derived units / have formulae as part of it? And/or call functions do so for it?

No, no, and no. The condition tags system is meant to be as simple as possible (JSON only, no dependencies, etc.) so that it can be moved into C++ and run at system generation time for microsecond-level performance impact. Any/all of the above is something that should be implemented as part of SystemBody instead, and the derived value stored/exposed as an attribute.

> As I said, for humans, "breathable" is a combination of surface pressure and oxygen content (and absence of other toxic gases; inert gases don't matter). So might be a bit more complex..

Simplifying the complex problem is fine. For the most part, condition tags will be invisible to the player, and there are a multitude of "hidden factors" which may influence *why* an atmosphere is breathable (terraforming efforts, generational genetic adaptations, cybernetic assistance, etc.) from a scientific/lore perspective.

Keep in mind that at current, a "breathable atmosphere" as the code considers it is an atmosphere with the magic number "atmosOxidizing" greater than a completely unspecified (and unscientific!) amount. Even when the model is improved, the player is unlikely to ever see the precise gas composition of an atmosphere.

If a "breathable atmosphere" is sufficiently multivariate or nonlinear such that it cannot be approximated as a two or three variable "box" on a graph, then I'd suggest it should be implemented as a computed quantity of a SystemBody and a new boolean (or scalar!) field `atmosBreathable` added for it instead.

mwerle replied:
Ok, well out of scope for this PR :)

That said, I think it might be interesting to have fully-modelled (or at least better approximated) atmospheres at some point. I remember a discussion a while back talking about light scattering and realistic colours that would require/benefit from this as well.

And it could open exploration missions looking for new planets with specific atmospheric compositions - either for colonisation or specific industrial/mining applications.

sturnclaw commented on data/libs/Economy/Industry.lua:
```diff
@@ -0,0 +1,262 @@
+---@type table<string, Economy.IndustryModifier>
+IndustryDef.modifiers = {}
+
+-- Amount of workforce required to operate this industry
+IndustryDef.workforce = 1
+
```
Workforce is currently completely unimplemented, as station populations are drawn completely randomly "out of a hat". I have nebulous plans for using the workforce value (aggregated across industries) to determine things like station immigration and goods consumption, or to limit industries from being present on smaller stations, but nothing fully developed at this time.

sturnclaw replied:
It will be present in the documentation pass, yes. I may strike it completely if I cannot find a compelling reason to use the workforce value; it may be simply out-of-scope for the depth of simulation that is appropriate for Pioneer.

sturnclaw commented on data/libs/Economy/Industry.lua:
```diff
@@ -0,0 +1,262 @@
+-- Amount of workforce required to operate this industry
+IndustryDef.workforce = 1
+
+local function parseModifierString(i, o, str)
+	local io, comm, effect, val = string.match(str, "([io]):(%g+)([%+%-])(%d+)")
+	local amt = effect == "-" and -tonumber(val) or tonumber(val)
+
+	if io == "i" then
+		i[comm] = (i[comm] or 0) + amt
+	else
+		o[comm] = (o[comm] or 0) + amt
+	end
+end
```
I agree, both the conditions and industries files need to be cleaned up and documented.

sturnclaw commented on src/galaxy/SystemBody.h:
```diff
@@ -118,7 +118,14 @@ class SystemBodyData {
+	// TODO: atmosphere hydrogen-nitrogen ratio from mass, carbon-oxygen ratio from atmosCarbon
+	// (approximate the distribution of nitrogen vs. hydrogen by using the mass of the body; < 5 Me predominantly nitrogen, > 7 Me predominantly hydrogen)
+	// this differentiates nitrogen-rich and hydrogen rich atmospheres, as well as (hydro)carbon-rich from oxygen-rich atmospheres
+	// thus, an earth-like atmosphere would be atmosOxidizing = 0.3, atmosCarbon <= 0.05
```
This is correct. You'll notice that the current semantic model has Earth's atmosphere as `atmosOxidizing = 1.0`... which if taken literally, would mean the atmosphere is 100% O2. The current model has absolutely no way to differentiate between a carbon-dioxide atmosphere and a pure oxygen atmosphere, vs. a nitrogen, methane, or hydrogen atmosphere, other than magic number constants which are used to generate a flavor text string.

This comment describes a proposed physically-based system to be implemented at a later time, which expresses a sample earthlike atmosphere as ~30% O2, < 5% CO2, and the remainder N2 and trace gasses.

mwerle replied:
Ok, sorry, I don't understand the current mechanism so I'll close this comment.

sturnclaw commented:
> We can make a commodity of every chemical element and base chemical compound

I appreciate the enthusiasm, but this is definitely not the direction we want to go. New commodities should be added because they fill a logical "gap" in a production chain or in commodity pricing, i.e. adding a "Chemicals" commodity produced at a refinery and consumed at a laboratory, or adding a high-value commodity to fill the gap between Robots and Precious Metals.

> Can be made, but even Hydrogen needs to be scooped and transported, and what about Venus and her CO2? => extended scoop equipment and even scoop missions?

Adding additional mission types, equipment, etc. is decidedly outside of the scope of this PR. It is the underlying assumption of this economic simulation that this sort of NPC trade is happening "in the background", and making these equilibrium activities visible to the player is a different topic.

> if it generates a starport in the orbit around Venus who processes CO2 from her atmosphere with H2 from the outer solar system to Carbon ore and plastics with O2 as byproduct, or even better imports H2O from ice moons to get the H2 from fuel cells because you can directly use electric current from more than 2000W/m2 solar radiation which is far more efficient than the 1340W/m2 we get in Earth's orbit. O2 may then be shipped back to the gas giants to be used in fuel cells, which could be another additional commodity. And oh yes, fuel cells are a source of energy to be tagged if no other source is available, meaning a demand for H2 and O2.

Condensing solid commodities from atmospheric gas is unlikely to be efficient enough to make a material difference to trade. Energy is not addressed by this economic simulation, as it is assumed that solar power and advanced microfusion using H2 provides more than enough power for industrial purposes.

For the most part, the complexity of simulation you describe above is not the focus of Pioneer. It can certainly be implied that all of this is happening in the background, carried out by the millions of people who inhabit star systems, but Pioneer is not a "tycoon" or economic-resource-management game and the player owning and managing a space station, etc. is a stated anti-goal of the project.

mwerle commented on data/economy/industries/agricultural.json:
```diff
@@ -0,0 +1,62 @@
+			"grain": 3
+		},
+		"modifiers": {
+			"rare_microbial_life": [ "i:fertilizer-1", "o:*+1" ]
```
Anywhere to list  "todo" items so things like this won't be forgotten/pushed down the track ad infinitum? (Happy to help writing docs once I understand it sufficiently..)

sturnclaw replied:
Feel free to add a comment in this PR with a list of things to document - both you and I will be able to edit that list. I plan on merging documentation with this PR, so as to prevent it from being pushed down the track :)

sturnclaw commented:
Here's the ~~very short~~ skeleton of the high-level design of this system for review purposes:

1. Conditions are a data-driven way for mods (and the base game!) to express named semantics about individual bodies. They are meant to eventually be determined at system generation time in C++. Because they have access to a random number generator, they can be used by more than just the economic system; for example, conditions can be used as a way to designate planets with valuable flora/fauna (for scanning missions), rare ores (for @impaktor's mining machinery) or any other kind of "planetary feature" developers/modders can think of.

2.  Industries express the economic endeavors of a (logical) subregion of a starport. For surface ports, this is a number of buildings, for orbitals it's a chunk of the ring. Starports have a certain number of industry "slots" based on their population. Workforce is an abstract logarithmic quantity meant to express the amount of population required to operate an industry, but currently has no interaction with any of the systems in this PR.

3. Each industry expresses its production and consumption in a dimensionless logarithmic rate quantity called "flow". The rules for flow are simple, but intentionally inconsistent. An industry which produces a commodity with flow N can supply one other industry at that station with demand flow N, and two or more industries each with demand flow N-1. Two or more industries with supply flow N can supply a single industry with demand flow N+1, but can *never* satisfy an industry with demand flow N+2. The precise base of the logarithm used in flow is intentionally not defined, but varies from 2 to 10+.

    This "rule quirk" regarding industries satisfying N+1 demand is meant to provide a meaningful (if mostly invisible) effect to randomly generating multiple industries which produce or consume the same commodity at the same rate. Its concrete effect is that the rare station which consumes 6 flow of ore *twice* will have a higher sell price for ore than a station which only has one industry which consumes 6 flow of ore.

6. Flow is converted to a physical quantity via the supply/demand rates. The natural logarithm is used to determine how many discrete units of a commodity the station produces and consumes over an intentionally undefined period of time. Supply/demand rates determine the trade price of a commodity at a station, and are affected by the market's `history` value which records the player's interactions with the market.

7. All stations are assumed (currently) to be at an implicit equilibrium. This means that all commodity demand will be satisfied, regardless of how much of that demand is actually produced at the station. This is modeled via two mechanics: local and interstellar trade. Local trade allows any surplus flow (flow that meets or exceeds indigenous demand) from a nearby station to be consumed by the station with a deficit, and is imported at lower prices. Interstellar trade is the assumed source of the remaining deficit not satisfied by local trade, and is imported at significantly higher prices. Both local and interstellar trade are currently assumed to be happening in the background, and the responsibility for "visualizing" this process by instantiating NPCs is left to other modules like TradeShips.

sturnclaw commented:
> To make suggestions for what you want: What is the wished and upper limit target quantity of industries and tradeable commodities? How are byproducts handled, means can industries have more than one commodity as output, more than two commodities as output, as rubbish as commodity yet exists?

The number of industry definitions has no hard limit. Avoiding too much granularity at this stage is desired (i.e. don't add a Mining Machinery Factory industry) as the industry system and number of discrete industries at a station are currently balanced around somewhat generic aggregations of multiple factories/companies/buildings into a logical "industry type".

As mentioned above, the number of commodities is something I'd like to avoid bloating, to keep each commodity feeling like it has an "identity" rather than disappearing into a sea of incredibly specific commodities. Additions for commodities should have a clear and defensible purpose.

Industries have any number of input and output commodities. Condition tags can be associated with modifiers to an industry's input and output commodities, which can increase and decrease commodity quantity and add new I/O commodities entirely.

sturnclaw committed:
economy: rename ConditionDef.conditions -> required

sturnclaw committed:
Update NaturalDocs options to 2.3.1

- Update CMakeLists.txt to invoke NaturalDocs with modern options.
- Update codedoc workflow to use options from ND project file, and use ND 2.3.1.
- Update gitignore for new NaturalDocs working directory.

sturnclaw committed:
Document FileSystem:ReadDirectory return values

sturnclaw committed:
economy: Document Conditions.lua

sturnclaw committed:
utils: to_array, keys_to_array

- Both functions can be used to generate an array from the key-part or value-part of a table.
- Filter predicate is optional on both functions.

sturnclaw committed:
Economy: document C++-side commodity and economy datastructures

- CPPEconomyInfo is likely to be deprecated once condition-based economies are hoisted to C++ code.
- CPPCommodityInfo will likely be read as raw JSON instead of a special structure.

sturnclaw committed:
Economy: document Economy.lua

sturnclaw committed:
Update copyright headers for new year

sturnclaw committed:
Economy: add the Chemicals commodity

- Meant to fill the gap between carbon ore and medicines, chemicals are produced from organic and inorganic sources.
- They replace grain as the input to a number of industries.
- Medicines and industrial equipment are made more expensive, and chemicals take the price point previously assigned to ind. equipment.

sturnclaw committed:
Fix condition file path missing data:// prefix

sturnclaw committed:
Rebalance and expand station industries to include chemical production

sturnclaw committed:
Document Industry.lua

sturnclaw committed:
Minor spelling fixes

sturnclaw committed:
Economy: remove unused industry workforce value

- This value was originally intended to drive station population based on industry, or to compute the commodity consumption of the population at the station, or to restrict industries based on the station's population.
- It has not been used and no concrete plan exists to make it relevant to the current state of starport generation.

sturnclaw committed:
Economy: add follow-on industry construction

- This allows modelling some level of intelligent design and planning when generating starport industries, allowing human authors to express relationships between industries.
- Only a single industry can be built from the list.
- The list of industries is evaluated every time an industry is selected, including for industries that were built as a follow-on  industry.
- Condition tags can be used to optionally filter the choice of industry.
- The percentage-based chance system provides variance between stations on the same body, keeping industry selection from appearing identical or guaranteed. A default chance of 50% is applied for each item if not specified.

sturnclaw committed:
Commodity: add GetProperName() accessor for capitalized commodity names

- Currently only works for Latin names due to string.upper's lack of support for UTF-8.
- Returns a commodity name capitalized to serve as a label or standalone list item.

sturnclaw committed:
PiGui: remove useless forwarding wrapper around sameLine()

sturnclaw committed:
ui: improve bookmark sidebar

- Better handling for commodity-route bookmarks.
- Add button to bookmark route destination as well as origin.
- Better sorting of bookmarks with routes.
- Fix error when displaying a commodity bookmark with an unknown commodity.

sturnclaw committed:
Economy: apply system-wide price modifier to station prices

sturnclaw committed:
ui: update system-econ-view for new economy

- Fix formatting of economic indicator icons.
- Normalize stylistic conventions with other sidebar widgets (use of framing, child window padding, etc.)
- Update debug mode to work with new economic simulation. Now displays equilibrium prices of stations in the selected system, caches station market data, and autoloads the current station's data.
- Debug mode does not display the current persistent market state at the given station. It only displays what the market state would be if initialized at the present time.

sturnclaw committed:
Economy: rebalance commodity prices slightly

- Metal alloys are more expensive per-ton than their ores.
- Battle weapons fit between slaves precious metals as the second-most expensive commodity per-ton.
- Textiles and fertilizer are more expensive per ton.
- Farm machinery was absurdly cheap per ton. Given it includes the catch-all of hydroponics and outdoor agricultural equipment, boosted to between industrial machinery and robots.
- These price changes are intended to reflect the production and consumption quantities used in the industry simulation. The lower the maximum production amount of a commodity, the more expensive it should be (with variance to account for desirability, monopoly, etc.)

sturnclaw committed:
Apply documentation review feedback

sturnclaw committed:
Add lui.VIEW_DESTINATION
